draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -722,6 +839,68 @@ Elligator2 can be simplified with projective coordinates.
 
+## FT-hash {#fthash}
+
+{{FT12}} outline a method for hashing onto a set of curves which includes BN curves, which are suitable
+for bilinear pairings. A number of cryptographic schemes rely on pairings, including some identity-based
+encryption (IBE) and attribute-based encryption (ABE) schemes.  BN curves are not one specific curve but a
+class of curves which can be constructed using Barreto and Naehrig's algorithm {{BN05}}. BN curves are a
+preferred option for asymmetric pairings as they provide relatively efficient computation for pairings, which are
+usually the slowest computation in any cryptosystem using them. There is also work giving parameter selection
+for 256-bit BN-curves, resulting in 128 bits of security for hashing onto BN curves. Suggestions for suitable 224,
+256, 384, and 512-bit curves which are compliant with the ISO document and are based on M-type twisted curve can
+be found in {{ISO09EC}}. Other work on parameter selection for suitable curves can be found in {{BGMO+10}}
+{{AKLG+11}} {{Scott13}} {{BCMN+16}}.
+
+This method is suitable for BN-curves, which have the form E: f(x) = y^2 = x^3 + b, where operations
+are over a field Fp,  where p = 7 (mod 12) and 1 + b is a nonzero square in Fp. Fouque and Tibouchi give an 
+encoding F which makes the hash function Hash(m) = F(H1(m)) + F(H2(m)) indifferentiable from a random oracle,
+where H1 and H2 are independent hash functions into Fp. We give this encoding and hash function construction
+in the steps below.
+
+~~~
+map2curve_BN(t)
+
+Input:
+
+  t - the input, an element in Fp
+
+  f() - Curve function
+
+Output:
+
+  (x, y) - a point in E
+
+Steps:
+
+1.   s = (-3)^(1/2) (mod p)
+2.   x1 = (-1 + s)/2  -  (s t^2) / (1 + b + t^2) (mod p)
+3.   x2 = (-1 - s)/2  +  (s t^2) / (1 + b + t^2) (mod p)
+4.   x3 = 1 - ((1 + b + t^2)^2)  /  (3 t^2) (mod p)
+5.   s1 = (f(x1) ^ ((p - 1) / 2))  // calculate quadratic residue of f(x1)
+6.   s2 = (f(x2) ^ ((p - 1) / 2))  // calculate quadratic residue of f(x2)
+7.   s3 = (f(x3) ^ ((p - 1) / 2))  // calculate quadratic residue of f(x3)
+8.   x = x3
+9.   x = CMOV(x2 ,x, s2)
+10.  x = CMOV(x1, x, s1)
+11.  y = f(x)^(1/2) (mod p)
+12.  y = (t ^ ((p -1) / 2)) * y (mod p)
+13.  Output (x,  y)
+~~~
+Note that s will always be an integer in the field, due our base assumptions about p. If it is necessary to
+specifically define a point for F(0), then a suitable suggestion is f(0) = ( (-1 + s)  /  2, (1+b)^(1/2) ).
+In steps 5-7 all relevant quadric residues are calculated in order to prevent possible timing attacks where
+s2 and s3 are not calculated if s1 = 1.
+
+Whilst this encoding function works for BN curves and is efficient, it only reaches roughly 9/16ths of
+the curve. However, this is enough to be a suitable encoding function for Hash = F(H1(m)) +
+F(H2(m)) as explained in {{#ffstv}}, which will be indifferentiable from a random oracle.
+This is discussed in in the Random Oracles section {#random-oracles}.
+
+# Curve Transformations
```

Comment:
I think this section was written already -- can you please check and remove if so?